R/enumerate_all_paths.R

```diff
@@ -0,0 +1,160 @@
+                                     former_key_cols = character(),
+                                     path = character(),
+                                     graph_df_ud,
+                                     helper_env) {
+  if (length(path) > 0) {
+    # increase tbl_node[[node]] by 1, return this index in a suffix
+    usage_idx <- inc_tbl_node(node, helper_env)
+    add_path_to_all_paths(
+      graph_df_ud,
+      node,
+      node_key_cols,
+      former_node,
+      former_key_cols,
+      new_former_node = names(path)[[length(path)]],
+      usage_idx,
+      helper_env
+    )
+  } else {
+    usage_idx <- ""
+  }
+
+  path <- c(path, set_names(node, paste0(node, usage_idx)))
+  out <-
+    graph_df_ud %>%
+    filter(child_table == !!node) %>%
+    filter(!(parent_table %in% path)) %>%
+    rename(node = parent_table, node_key_cols = parent_cols, former_node = child_table, former_key_cols = child_cols) %>%
+    select(-on_delete)
+
+  if (nrow(out) == 0) {
+    return()
+  }
+
+  pwalk(out, enumerate_all_paths_impl, path, graph_df_ud, helper_env)
+}
+
+rename_unique <- function(all_paths) {
+  all_names <- bind_rows(
+    select(all_paths, table = child_table, new_table = new_child_table),
+    select(all_paths, table = parent_table, new_table = new_parent_table)
+  ) %>%
+    distinct() %>%
+    arrange(table, new_table) %>%
+    group_by(table) %>%
+    count()
+  left_join(all_paths, rename(all_names, n_c = n), by = c("child_table" = "table")) %>%
+    left_join(rename(all_names, n_p = n), by = c("parent_table" = "table")) %>%
+    mutate(
+      new_child_table = if_else(n_c == 1, child_table, new_child_table),
+      new_parent_table = if_else(n_p == 1, parent_table, new_parent_table)
+    ) %>%
+    select(-n_c, -n_p)
+}
+
+inc_tbl_node <- function(node, helper_env) {
+  tbl_node <- get("tbl_node", envir = helper_env)
+  tbl_node[[node]] <- (tbl_node[[node]] %||% 0) + 1
+  assign("tbl_node", tbl_node, envir = helper_env)
+  paste0("-", tbl_node[[node]])
+}
+
+add_path_to_all_paths <- function(graph_df_ud,
+                                  node,
+                                  node_key_cols,
+                                  former_node,
+                                  former_key_cols,
+                                  new_former_node,
+                                  usage_idx,
+                                  helper_env) {
+  all_paths <- get("all_paths", helper_env)
+  path_element <- graph_df_ud %>% filter(
+    (
+      child_table == node &
+        map_lgl(child_cols, ~ identical(sort(.x), sort(node_key_cols))) &
+        parent_table == former_node &
+        map_lgl(parent_cols, ~ identical(sort(.x), sort(former_key_cols)))
+    ) |
+      (
+        parent_table == node &
+          map_lgl(parent_cols, ~ identical(sort(.x), sort(node_key_cols))) &
+          child_table == former_node &
+          map_lgl(child_cols, ~ identical(sort(.x), sort(former_key_cols)))
+      )
```

Comment:
Can we simplify this (and similarly below)? In `graph_df_ud`, each original edge in the data model graph should correspond to two rows, one for each direction.

Reply:
Would be great if we can make it simpler. I am still wondering though:
Here I am searching for the connection that's really existing and subsequently add it to the result table. 
Since we arrived here via an edge in the undirected graph, we do not know if node is the child table or the parent table. I figured I need to filter for both possibilities. In the code below I then take the first hit, since that is the correct one (since we earlier did `bind_rows(correct_graph_tbl, inverse_graph_tbl)`)

It is not enough to store the sequence of tables in the result, since in cases like e.g. `dm_nycflights13()` we then do not know which FK we took from `flights` to `airports`.
It is also not enough to just filter:
```r
filter(
	child_table %in% c(node, former_node),
	map_lgl(child_cols, ~ identical(sort(.x), sort(node_key_cols)) | identical(sort(.x), sort(former_key_cols))),
	parent_table %in% c(node, former_node),
	map_lgl(parent_cols, ~ identical(sort(.x), sort(node_key_cols)) | identical(sort(.x), sort(former_key_cols)))
	)
```
since then we could run into problems with ambiguities.

Maybe I am missing something.

Reply:
This might be easier if the foreign keys were named. That shouldn't become a blocker, though.

I'll take a look.